src/ucp/api/ucp.h

```diff
@@ -3033,6 +3253,20 @@ void ucp_stream_data_release(ucp_ep_h ep, void *data);
 void ucp_request_free(void *request);
 
 
+/**
+ * @ingroup UCP_COMM
+ * @brief Create an empty communications request.
+ *
+ * @return Error code as defined by @ref ucs_status_t
+ *
+ * This routine creates request which may be used in functions
+ * @ref ucp_tag_send_nbx, @ref ucp_tag_recv_nbx, etc. The application
+ * is responsible for releasing the handle using the @ref ucp_request_free
+ * routine
+ */
+void *ucp_request_alloc();
```

Comment:
do you need to provide worker as a parameter? Would be good to get request from worker memory pool

Reply:
Should there be an `nreqs` parameter as well to allocate an array of requests?

Reply:
> Should there be an `nreqs` parameter as well to allocate an array of requests?

then need to add another parameter - buffer where UCP would put request pointers... Would not be better to allocate array of requests (instead of pointers) by user? the size can be queried by `ucp_context_query`

Reply:
UCX does not hold the requests as a contiguous block, so allocating an array is not much more optimal that allocating one by one.
What is the use case to allocate an array?

Reply:
pass worker_h param

Reply:
added worker argument

Reply:
>  allocating an array is not much more optimal that allocating one by one

If this is the case, then I agree there's no benefit.